full-stack-tests/presentation/src/learning-snippets/Hierarchies.test.ts

```diff
@@ -24,305 +28,1618 @@ describe("Learning Snippets", () => {
+        // by label.
+        const ruleset: Ruleset = {
+          id: "example",
+          rules: [{
+            ruleType: RuleTypes.RootNodes,
             specifications: [{
-              specType: ChildNodeSpecificationTypes.CustomNode,
-              type: "A",
-              label: "A",
+              specType: ChildNodeSpecificationTypes.InstanceNodesOfSpecificClasses,
+              classes: [{ schemaName: "ECDbMeta", classNames: ["ECPropertyDef"] }],
+              groupByLabel: false,
+            }],
+          }],
+        };
+        // __PUBLISH_EXTRACT_END__
+        printRuleset(ruleset);
+
+        // Verify that instances were not grouped by label
+        const classGroupingNodes = await Presentation.presentation.getNodes({ imodel, rulesetOrId: ruleset });
+        const classGroupingNode = classGroupingNodes.find((node) => {
+          assert(NodeKey.isClassGroupingNodeKey(node.key));
+          return node.key.className === "ECDbMeta:ECPropertyDef";
+        })!;
+        const nodes = await Presentation.presentation.getNodes({ imodel, rulesetOrId: ruleset, parentKey: classGroupingNode.key });
+        expect(nodes).to.not.be.empty;
+        nodes.forEach((node) => expect(NodeKey.isLabelGroupingNodeKey(node.key)).to.be.false);
+      });
+
+      it("uses `hasChildren` attribute", async () => {
+        // __PUBLISH_EXTRACT_START__ Hierarchies.Specification.HasChildren.Ruleset
+        // The ruleset contains a root nodes' specification that returns `bis.Model` nodes and a child nodes'
+        // specification that returns a custom node. The `bis.Model` nodes would have the custom node as a
+        // child, but `hasChildren: "Never"` attribute overrides that.
+        const ruleset: Ruleset = {
+          id: "example",
+          rules: [{
+            ruleType: RuleTypes.RootNodes,
+            specifications: [{
+              specType: ChildNodeSpecificationTypes.InstanceNodesOfSpecificClasses,
+              classes: [{ schemaName: "BisCore", classNames: ["Model"], arePolymorphic: true }],
+              hasChildren: "Never",
             }],
           }, {
-            condition: `FALSE`,
+            ruleType: RuleTypes.ChildNodes,
             specifications: [{
               specType: ChildNodeSpecificationTypes.CustomNode,
-              type: "B",
-              label: "B",
+              type: "T_CHILD",
+              label: "Child",
             }],
           }],
-        }],
-      };
-      // __PUBLISH_EXTRACT_END__
-
-      // The root node rule meets schema requirement, but only the first sub-condition's condition
-      // attribute evaluates to `true` - expect only the "A" node.
-      const nodes = await Presentation.presentation.getNodes({ imodel, rulesetOrId: ruleset });
-      expect(nodes.length).to.eq(1);
-      expect(nodes).to.containSubset([{
-        label: { displayValue: "A" },
-      }]);
+        };
+        // __PUBLISH_EXTRACT_END__
```

Comment:
I think it would have been more appropriate to set `hasChildren` to `Always` to demonstrate the correct usage of this attribute. The unexpected results that may be encountered by incorrect usage have already been covered by the paragraph above.

Reply:
changed the example, added screenshots

Reply:
The screenshots are very helpful, but the example still demonstrates the misuse of the attribute.

Reply:
Oh, I now understand what you're saying. Setting it to `Always` when it has children or `Never` when it doesn't makes it impossible to show that it works through a screenshot or a code snippet.

Reply:
No need to corner yourself with the pattern of your creation. This particular attribute is weird, so naturally it does not fit into the same mold as the rest.

If you really want to "prove" that your example works (which I argue you shouldn't), you could show the difference in tree load time when `hasChildren` is set.

Reply:
I'd like to keep it as is based on the following:
- Showing performance changes is challenging:
  - We'd need to run the test on a large iModel to make the difference noticeable.
  - Simply hardcoding some performance numbers wouldn't look good, IMO.
- Some example is better than no example

Reply:
I never suggested to remove the example, just change it to showcase the correct usage of this attribute.

Reply:
You mean showing just the rules, right? How does that help if we don't show the effect of the attribute?

Reply:
Thanks for addressing my concerns.